File: plugin/src/main/java/org/octopusden/octopus/jira/api/config/ApiSettingsProvider.java
package org.octopusden.octopus.jira.api.config;

import com.atlassian.sal.api.pluginsettings.PluginSettings;
import com.atlassian.sal.api.pluginsettings.PluginSettingsFactory;
import org.octopusden.octopus.jira.exception.JiraApplicationException;

import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

public class ApiSettingsProvider {

    private final PluginSettings pluginSettings;
    private final Properties properties;

    public ApiSettingsProvider(PluginSettingsFactory pluginSettingsFactory) {
        this.pluginSettings = pluginSettingsFactory.createGlobalSettings();
        this.properties = loadProperties();
    }

    private Properties loadProperties() {
        Properties props = new Properties();
        try (InputStream stream = ApiSettingsProvider.class.getResourceAsStream("/octopus-jira-api-plugin.properties")) {
            props.load(stream);
        } catch (IOException e) {
            throw new RuntimeException("Failed to load properties file", e);
        }
        return props;
    }

    public Object get(ApiSetting setting) {
        Object value = pluginSettings.get(setting.getKey());
        if (value != null) {
            return value;
        }
        return getProperty(setting.getKey());
    }

    public void set(ApiSetting setting, String value) {
        pluginSettings.put(setting.getKey(), value);
    }

    public String getString(ApiSetting setting) {
        Object value = get(setting);
        if (value instanceof String) {
            return (String) value;
        }
        throw new IllegalStateException("Setting '" + setting.getKey() + "' is not configured");
    }

    private String getProperty(String propertyName) {
        String value = properties.getProperty(propertyName);
        if (value == null) {
            throw new JiraApplicationException(
                    "Property [" + propertyName + "] must be defined in the property file [/octopus-jira-api-plugin.properties]"
            );
        }
        return value;
    }

    public Properties getProperties() {
        return properties;
    }
}

File: plugin/src/main/kotlin/org/octopusden/octopus/jira/api/config/ApiSettingsProvider.kt
package org.octopusden.octopus.jira.api.config

import com.atlassian.sal.api.pluginsettings.PluginSettingsFactory
import org.octopusden.octopus.jira.exception.JiraApplicationException
import java.util.Properties

private const val PROPERTIES_FILE_NAME = "/octopus-jira-api-plugin.properties"

class ApiSettingsProvider(pluginSettingsFactory: PluginSettingsFactory) {

    private val pluginSettings = pluginSettingsFactory.createGlobalSettings()

    val properties = Properties().apply {
        ApiSettingsProvider::class.java.getResourceAsStream(PROPERTIES_FILE_NAME)
            .use { stream ->
                load(stream)
            }
    }

    operator fun get(setting: ApiSetting): Any =
        pluginSettings[setting.key] ?: getProperty(setting.key)

    operator fun set(setting: ApiSetting, value: String?) {
        pluginSettings.put(setting.key, value)
    }

    private fun getProperty(propertyName: String): String {
        return properties.getProperty(propertyName)
            ?: throw JiraApplicationException("Property [$propertyName] must be defined in the property file [$PROPERTIES_FILE_NAME]")
    }

    fun getString(key: ApiSetting): String =
        get(key) as? String
            ?: throw IllegalStateException("Setting '${key.key}' is not configured")
}

enum class ApiSetting(
    val key: String,
    val displayName: String,
    val type: SettingType,
    val values: List<String> = emptyList()
) {
    IPS_REPORTS_PROJECT("ips-reports.project", "IPS Reports Project", SettingType.TEXT),
    SERVICE_USER("service-user.username", "Service User", SettingType.TEXT);

    enum class SettingType { TEXT, PASSWORD, SELECT }
}
